### PR TITLE
Update Kubernetes to 1.27.2 and Kustomize to 5.0.3

### PR DESCRIPTION
**Pull request by aryan9600**
* Bump all k8s.io packages to be up to date with Kubernetes 1.27.2
* Bump sigs.k8s.io/kustomize packages to be up to date with Kustomize 5.0.3
* Bump sigs.k8s.io/controller-runtime to `v0.15.0-beta.0` 
* Add a new util func `NewDynamicRESTMapper(restConfig *rest.Config)` in `runtime/client` for creating new RESTMapper objects.

Part of: https://github.com/fluxcd/flux2/issues/3564

### CODE REVIEW

File: kustomize/go.mod
Comment (stefanprodan) [inline kustomize/go.mod:26]: Please restore this and update the version in the comment, this is used to tell Flux users what Kustomize version we're using, as the api & kyaml versions are confusing.

**Comment by hiddeco** [comment]
Converted to draft until the controller-runtime fork is replaced with official release.

File: runtime/client/impersonator.go
Comment (stefanprodan) [inline runtime/client/impersonator.go:139]: Please create a function in runtime/client/client.go named `NewDynamicRESTMapper` that wraps this code. Then in all our controllers we can use this helper.

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @aryan9600 🎖️

File: ssa/go.mod
Comment (stefanprodan) [inline ssa/go.mod:7]: This will cause major issues downstream as every consumer must also use a replacement. Please set the controller-runtime to v0.13.1-0.20230504051641-57674c8dd62f in the require section.

**Review by stefanprodan — CHANGES_REQUESTED** [review]

**Comment by aryan9600** [comment]
Didn't notice that package as it had no direct deps to Kubernetes or Kustomize packages 😓 Let me update it as well.

**Comment by stefanprodan** [comment]
We decided to wait for controller-runtime v0.15 release that is blocked by a regression in Kubernetes v1.27 that will be fixed in Kubernetes 1.27.2 (scheduled for 17th May). 

Ref: https://github.com/kubernetes-sigs/controller-runtime/issues/2297

**Review by hiddeco — COMMENTED** [review]
Can you update the controller-runtime dependencies to the `v0.15.0-beta.0` tag?

After discussion with @stefanprodan, we have decided to bet on this for lower level dependencies, while allowing the controllers to manually upgrade to the final release. 

This would unblock the controllers from prepping in a sane way.

**Pull request merged**
